<a href="https://colab.research.google.com/github/gerardowarbringer-eng/Programacion-para-Analitica-Descrip-Predict/blob/main/Actividad_Calificada_07_Datos_Duplicados_271770.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Actividad calificada: Datos duplicados**
### Detección, cuantificación y tratamiento de duplicados en un dataset real

---

| | |
|---|---|
| **Materia** | Programación para Analítica Descriptiva y Predictiva |
| **Programa** | Maestría en Inteligencia Artificial y Analítica de Datos (MIAAD) — UACJ |
| **Unidad** | 02 Análisis Descriptivo de los Datos |
| **Tema** | Limpieza y transformación de datos: duplicados |
| **Modalidad** | Individual |
| **Valor** | 100 puntos |
| **Dataset** | Hotel Booking Demand (Antonio, Almeida y Nunes, 2019) |
| **Fecha de entrega** | [completar] |

---

## Datos del estudiante

| | |
|---|---|
| **Nombre completo** | GERARDO ISIDORO REYES|
| **Matrícula** | 271770 |

## Instrucciones

1. Trabaja en Google Colab. Guarda una copia en tu Drive (**Archivo → Guardar una copia en Drive**) antes de empezar.
2. Completa tus datos en la tabla anterior.
3. Resuelve las actividades **en orden**: algunas usan variables creadas en actividades anteriores.
4. Escribe tu código en las celdas marcadas con `# Escribe tu código aquí`. Puedes agregar celdas si lo necesitas.
5. **Comenta tu código**: cada bloque debe explicar qué hace y por qué. Los comentarios forman parte de la calificación de cada actividad.
6. Responde las preguntas en las celdas **Respuesta**, con tus propias palabras y apoyándote en los resultados que obtuviste (cita cifras).
7. Usa los métodos vistos en el notebook de la sesión: `duplicated()`, `drop_duplicates()`, `value_counts()`, `sort_values()`, `groupby()`, `pd.concat()`, máscaras booleanas, etc.
8. Antes de entregar, ejecuta todo el notebook (**Entorno de ejecución → Reiniciar y ejecutar todo**) y verifica que no haya errores.
9. **Entrega en Moodle** un archivo de texto `Actividad_Duplicados_<Matricula>.txt` que contenga el **enlace** a tu notebook (Google Colab o GitHub), con las salidas visibles.
   - Colab: **Compartir → Acceso general → Cualquier persona con el enlace (Lector)**.
   - GitHub: el repositorio debe ser **público**.
   - Opcional: puedes adjuntar también el archivo `.ipynb`.

## Contexto

El dataset **Hotel Booking Demand** contiene 119,390 reservas de dos hoteles de Portugal (un hotel urbano, *City Hotel*, y un hotel vacacional, *Resort Hotel*) con llegada entre julio de 2015 y agosto de 2017. Cada fila es una reserva y hay 32 columnas. Algunas relevantes para esta actividad:

| Columna | Descripción |
|---|---|
| `hotel` | `City Hotel` o `Resort Hotel` |
| `is_canceled` | 1 si la reserva se canceló, 0 si no |
| `lead_time` | Días entre la reserva y la fecha de llegada |
| `arrival_date_year`, `arrival_date_month`, `arrival_date_day_of_month` | Fecha de llegada |
| `adults`, `children`, `babies` | Número de huéspedes |
| `country` | País de origen (código ISO de 3 letras) |
| `market_segment` | Segmento de mercado (`Online TA`, `Groups`, `Direct`, …) |
| `deposit_type` | Tipo de depósito (`No Deposit`, `Non Refund`, `Refundable`) |
| `agent`, `company` | Identificador de la agencia o empresa que hizo la reserva (con faltantes) |
| `customer_type` | Tipo de cliente (`Transient`, `Group`, …) |
| `adr` | Tarifa diaria promedio (*average daily rate*) |
| `reservation_status` | Estado final: `Check-Out`, `Canceled`, `No-Show` |
| `reservation_status_date` | Fecha del último cambio de estado |

> **Importante**: el dataset **no incluye un identificador de reserva**. Tenlo presente al interpretar los duplicados.

## Rúbrica

| Actividad | Puntos | Excelente | Bueno | Insuficiente | No entregado / Incorrecto |
|---|---|---|---|---|---|
| **1. Diagnóstico inicial** | 20 | 18–20: conteos y porcentajes correctos con los tres valores de `keep`; duplicados por hotel calculados; código comentado; conclusión con cifras. | 14–17: resultados correctos con omisiones menores en comentarios o en la conclusión. | 1–13: conteos incompletos o con errores; conclusión sin sustento. | 0: no se entregó o el código no produce resultados. |
| **2. Inspección y criterio** | 25 | 23–25: grupos inspeccionados y ordenados; tamaños de grupo analizados; comparación duplicados vs. total; argumento sólido sobre si son errores, con evidencia. | 18–22: análisis correcto pero con argumento parcial o poca evidencia. | 1–17: inspección superficial; argumento sin datos. | 0: no se entregó o es incorrecto. |
| **3. Impacto en el análisis** | 25 | 23–25: métricas por hotel antes y después con `groupby()`; tabla comparativa con `pd.concat()`; interpretación del efecto en el análisis y en un modelo predictivo. | 18–22: métricas correctas; tabla o interpretación incompletas. | 1–17: métricas con errores o sin comparación. | 0: no se entregó o es incorrecto. |
| **4. Tratamiento y verificación** | 20 | 18–20: limpieza verificada; duplicados parciales detectados e interpretados; registro de decisión completo. | 14–17: limpieza correcta; análisis parcial o registro incompleto. | 1–13: limpieza sin verificación o análisis parcial erróneo. | 0: no se entregó o es incorrecto. |
| **Reflexión final** | 10 | 9–10: respuestas fundamentadas en los resultados propios. | 7–8: respuestas correctas pero generales. | 1–6: respuestas vagas o sin relación con los resultados. | 0: no se entregó. |
| **Total** | **100** | | | | |

**Notas**:

- Las respuestas escritas de cada actividad (1.4, 2.4, 3.4, 4.2 y 4.3) y los comentarios del código cuentan para la calificación **de esa actividad**, no solo para la reflexión final.
- Entrega tardía: se descuenta **5 % por día natural** de retraso, hasta 7 días; después de ese plazo la calificación es 0. El descuento se aplica sobre la calificación obtenida.

## Carga de datos

Ejecuta la siguiente celda. No requiere modificaciones.

In [77]:
import pandas as pd

# Carga del dataset Hotel Booking Demand desde GitHub (TidyTuesday, 2020-02-11)
url = 'https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-02-11/hotels.csv'
hoteles = pd.read_csv(url)

# Dimensiones y primeras filas
print('Dimensiones:', hoteles.shape)
hoteles.head()

Dimensiones: (119390, 32)


,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,deposit_type,agent,company,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date
0,Resort Hotel,0,342,2015,July,27,1,0,0,2,...,No Deposit,NaN,NaN,0,Transient,0.0,0,0,Check-Out,2015-07-01
1,Resort Hotel,0,737,2015,July,27,1,0,0,2,...,No Deposit,NaN,NaN,0,Transient,0.0,0,0,Check-Out,2015-07-01
2,Resort Hotel,0,7,2015,July,27,1,0,1,1,...,No Deposit,NaN,NaN,0,Transient,75.0,0,0,Check-Out,2015-07-02
3,Resort Hotel,0,13,2015,July,27,1,0,1,1,...,No Deposit,304.0,NaN,0,Transient,75.0,0,0,Check-Out,2015-07-02
4,Resort Hotel,0,14,2015,July,27,1,0,2,2,...,No Deposit,240.0,NaN,0,Transient,98.0,0,1,Check-Out,2015-07-03


---
## Actividad 1. Diagnóstico inicial (20 puntos)

### 1.1 Conteo de duplicados exactos (5 pts)

Calcula el número de filas marcadas como duplicadas en `hoteles` considerando **todas** las columnas, con `keep='first'`, `keep='last'` y `keep=False`. Imprime los tres resultados con un texto que explique qué significa cada uno.

In [78]:
# Escribe tu código aquí
#Primero realizamos una copia del data frame:
comparacion = hoteles.copy()
# Conteo de duplicados

duplicados_first = comparacion.duplicated(keep='first').sum() #marca los datos repetidos excepto el primero
duplicados_last = comparacion.duplicated(keep='last').sum() #marca los datos repetidos excepto el último
duplicados_false = comparacion.duplicated(keep=False).sum() #marca todos los duplicados incluidos primero y utlimos.

print(f"keep='first': {duplicados_first} filas marcadas como duplicadas.")
print(f"keep='last': {duplicados_last} filas marcadas como duplicadas.")
print(f"keep=False: {duplicados_false} filas marcadas como duplicadas.")


keep='first': 31994 filas marcadas como duplicadas.
keep='last': 31994 filas marcadas como duplicadas.
keep=False: 40165 filas marcadas como duplicadas.


### 1.2 Magnitud del problema (5 pts)

Con los resultados de 1.1, calcula e imprime:

- El **porcentaje** de filas sobrantes respecto al total.
- El número de **grupos repetidos** (combinaciones de valores que aparecen más de una vez).

In [79]:
# Escribe tu código aquí
#Ignora la primera aparición de cada valor y sumando las repeticiones del df.
duplicados = comparacion.duplicated(keep='first').sum()

print(f"Porcentaje de filas sobrantes: {duplicados / len(comparacion) * 100:.2f}%")

grupos_repetidos = comparacion.duplicated(keep=False).sum()
print(f"Filas pertenecientes a grupos repetidos: {grupos_repetidos}")


Porcentaje de filas sobrantes: 26.80%
Filas pertenecientes a grupos repetidos: 40165


### 1.3 Duplicados por hotel (5 pts)

Determina **cuántas** filas sobrantes (`keep='first'`) hay en cada hotel y qué **porcentaje** representan respecto al total de reservas de ese hotel.

Sugerencia: crea una copia de `hoteles`, agrega una columna booleana `es_duplicado` con el resultado de `duplicated()` y agrupa por `hotel`.

In [80]:
# Escribe tu código aquí
# Crear una copia
comparacion = hoteles.copy()

# Marcar los duplicados exactos, conservando con "first"
comparacion['es_duplicado'] = comparacion.duplicated(keep='first')

# Agrupa por hotel
duplicados_por_hotel = comparacion.groupby('hotel').agg(
    filas_sobrantes=('es_duplicado', 'sum'),
    total_reservas=('hotel', 'size')
)

# Calcula el porcentaje de filas sobrantes
duplicados_por_hotel['porcentaje_sobrantes'] = (
    duplicados_por_hotel['filas_sobrantes'] /
    duplicados_por_hotel['total_reservas'] * 100
)

# Mostrar resultados
print(duplicados_por_hotel)


              filas_sobrantes  total_reservas  porcentaje_sobrantes
hotel                                                              
City Hotel              25902           79330             32.650952
Resort Hotel             6092           40060             15.207189


### 1.4 Conclusión (5 pts)

¿Qué tan grave es el problema de duplicados en este dataset? ¿Afecta por igual a ambos hoteles? Responde con cifras.

**Respuesta 1.4:**

_Escribe aquí tu respuesta._ Al observar los resultados de los duplicados, se concluye que representan una parte considerable del dataset y la proporción es claramente diferente entre los dos hoteles. En cuanto a la segunda pregunta; City Hotel reprenta el 32.65% y Resort Hotel un 15.20% que es el doble de duplicados en cuanto a porcentaje se refiere.

---
## Actividad 2. Inspección y criterio de duplicado (25 puntos)

### 2.1 Ver los grupos repetidos (6 pts)

Filtra las filas involucradas en duplicados (`keep=False`), **ordénalas** con `sort_values()` para que las copias queden juntas y muestra las primeras 12 filas con estas columnas:

`hotel`, `lead_time`, `arrival_date_year`, `arrival_date_month`, `arrival_date_day_of_month`, `adults`, `country`, `market_segment`, `deposit_type`, `adr`, `reservation_status`.

Elige tú las columnas de ordenamiento y explica tu elección en un comentario.

In [81]:
# Escribe tu código aquí
# Filtrar todas las filas que pertenecen a un grupo de duplicados.
duplicados = comparacion[comparacion.duplicated(keep=False)].copy()

# Primero ordenamos por hotel y por las columnas que definen la reserva.
duplicados = duplicados.sort_values(
    by=[
        'hotel',
        'lead_time',
        'arrival_date_year',
        'arrival_date_month',
        'arrival_date_day_of_month',
        'adults',
        'country',
        'market_segment',
        'deposit_type',
        'adr',
        'reservation_status'
    ]
)

# Muest las primeras 12 filas con las columnas requeridas
duplicados[[
    'hotel',
    'lead_time',
    'arrival_date_year',
    'arrival_date_month',
    'arrival_date_day_of_month',
    'adults',
    'country',
    'market_segment',
    'deposit_type',
    'adr',
    'reservation_status'
]].head(12)


,hotel,lead_time,arrival_date_year,arrival_date_month,arrival_date_day_of_month,adults,country,market_segment,deposit_type,adr,reservation_status
76793,City Hotel,0,2015,August,10,2,FRA,Online TA,No Deposit,75.0,Check-Out
76794,City Hotel,0,2015,August,10,2,FRA,Online TA,No Deposit,75.0,Check-Out
41178,City Hotel,0,2015,August,14,2,PRT,Offline TA/TO,No Deposit,109.0,Canceled
41197,City Hotel,0,2015,August,14,2,PRT,Offline TA/TO,No Deposit,109.0,Canceled
41198,City Hotel,0,2015,August,14,2,PRT,Offline TA/TO,No Deposit,109.0,Canceled
41199,City Hotel,0,2015,August,14,2,PRT,Offline TA/TO,No Deposit,109.0,Canceled
41205,City Hotel,0,2015,August,14,2,PRT,Offline TA/TO,No Deposit,109.0,Canceled
41206,City Hotel,0,2015,August,14,2,PRT,Offline TA/TO,No Deposit,109.0,Canceled
41207,City Hotel,0,2015,August,14,2,PRT,Offline TA/TO,No Deposit,109.0,Canceled
41208,City Hotel,0,2015,August,14,2,PRT,Offline TA/TO,No Deposit,109.0,Canceled


### 2.2 Tamaño de los grupos (6 pts)

Usa `value_counts()` sobre el `DataFrame` completo para contar cuántas veces aparece cada combinación de filas. Considera que las columnas `agent` y `company` tienen faltantes (revisa el parámetro `dropna`).

Con ese resultado:

1. Muestra **cuántos grupos** hay de cada tamaño (2 copias, 3 copias, …). Basta con los 10 tamaños más frecuentes.
2. Indica el **tamaño del grupo más grande** y muestra su combinación de valores.

In [66]:
# Escribe tu código aquí

#1. Muestra cuántos grupos hay de cada tamaño (2 copias, 3 copias, …). Basta con los 10 tamaños más frecuentes.
# Contar cuántas veces aparece cada combinación exacta de columnas.
conteo_filas = comparacion.value_counts(dropna=False)

# Eliminar las combinaciones que aparecen una sola vez
grupos_repetidos = conteo_filas[conteo_filas > 1]

# con value counts cuenta los grupos hay de cada tamaño
grupos_por_tamano = grupos_repetidos.value_counts().sort_index()

# Muestra los 10 tamaños de grupo más frecuentes
print("10 tamaños de grupo más frecuentes:")
print(grupos_por_tamano.sort_values(ascending=False).head(10))


10 tamaños de grupo más frecuentes:
count
2     969
3     497
4     275
5     171
9     123
6     115
7     103
19     97
8      93
14     64
Name: count, dtype: int64


In [67]:
#2. Indica el tamaño del grupo más grande y muestra su combinación de valores.
# Tamaño máximo del grupo
tamano_maximo = conteo_filas.max()

print(f"\nTamaño del grupo más grande: {tamano_maximo} copias")

# Muestra la combinación de valores que forma del grupo
grupo_mas_grande = conteo_filas[conteo_filas == tamano_maximo]

print("\nCombinación de valores del grupo más grande:")
print(grupo_mas_grande)



Tamaño del grupo más grande: 179 copias

Combinación de valores del grupo más grande:
hotel       is_canceled  lead_time  arrival_date_year  arrival_date_month  arrival_date_week_number  arrival_date_day_of_month  stays_in_weekend_nights  stays_in_week_nights  adults  children  babies  meal  country  market_segment  distribution_channel  is_repeated_guest  previous_cancellations  previous_bookings_not_canceled  reserved_room_type  assigned_room_type  booking_changes  deposit_type  agent  company  days_in_waiting_list  customer_type  adr    required_car_parking_spaces  total_of_special_requests  reservation_status  reservation_status_date  es_duplicado
City Hotel  1            277        2016               November            46                        7                          1                        2                     2       0.0       0       BB    PRT      Groups          TA/TO                 0                  0                       0                               A         

### 2.3 ¿En qué se distinguen las filas duplicadas? (6 pts)

Compara las filas involucradas en duplicados (`keep=False`) con el dataset completo en:

- Distribución de `market_segment` (proporciones).
- Distribución de `deposit_type` (proporciones).
- Tasa de cancelación (`is_canceled`).

In [68]:
# Escribe tu código aquí
# Distribución de market_segment en las filas duplicadas
distribucion_market_segment = (
    comparacion[comparacion.duplicated(keep=False)]['market_segment']
    .value_counts(normalize=True)
)

print(distribucion_market_segment)


market_segment
Groups           0.529154
Offline TA/TO    0.346884
Online TA        0.081231
Corporate        0.030675
Direct           0.011536
Complementary    0.000445
Aviation         0.000074
Name: proportion, dtype: float64


In [69]:
# Dataset completo
distribucion_market_segment_total = hoteles['market_segment'].value_counts(normalize=True)

# Filas duplicadas
distribucion_market_segment_duplicados = (
    comparacion[comparacion.duplicated(keep=False)]['market_segment']
    .value_counts(normalize=True)
)

print("Dataset completo:")
print(distribucion_market_segment_total)

print("\nFilas duplicadas:")
print(distribucion_market_segment_duplicados)


Dataset completo:
market_segment
Online TA        0.473046
Offline TA/TO    0.202856
Groups           0.165935
Direct           0.105587
Corporate        0.044350
Complementary    0.006223
Aviation         0.001985
Undefined        0.000017
Name: proportion, dtype: float64

Filas duplicadas:
market_segment
Groups           0.529154
Offline TA/TO    0.346884
Online TA        0.081231
Corporate        0.030675
Direct           0.011536
Complementary    0.000445
Aviation         0.000074
Name: proportion, dtype: float64


In [70]:
# Distribución de deposit_type
distribucion_deposit_total = hoteles['deposit_type'].value_counts(normalize=True)

distribucion_deposit_duplicados = (
    comparacion[comparacion.duplicated(keep=False)]['deposit_type']
    .value_counts(normalize=True)
)

print("Deposit type - dataset completo:")
print(distribucion_deposit_total)

print("\nDeposit type - filas duplicadas:")
print(distribucion_deposit_duplicados)


Deposit type - dataset completo:
deposit_type
No Deposit    0.876464
Non Refund    0.122179
Refundable    0.001357
Name: proportion, dtype: float64

Deposit type - filas duplicadas:
deposit_type
No Deposit    0.499407
Non Refund    0.498850
Refundable    0.001743
Name: proportion, dtype: float64


In [71]:
# Tasa de cancelación
tasa_cancelacion_total = hoteles['is_canceled'].mean()

tasa_cancelacion_duplicados = (
    comparacion[comparacion.duplicated(keep=False)]['is_canceled']
    .mean()
)

print(f"Tasa de cancelación - dataset completo: {tasa_cancelacion_total:.2%}")
print(f"Tasa de cancelación - filas duplicadas: {tasa_cancelacion_duplicados:.2%}")


Tasa de cancelación - dataset completo: 37.04%
Tasa de cancelación - filas duplicadas: 68.54%


### 2.4 Criterio (7 pts)

El dataset no tiene identificador de reserva. Con base en la evidencia de 2.1, 2.2 y 2.3, responde:

a) ¿Las filas duplicadas son **errores de registro** o podrían ser **reservas distintas** que coinciden en todos sus valores? Argumenta con datos.

b) Las columnas `agent` y `company` tienen muchos faltantes. ¿Cómo trata `duplicated()` esos `NaN` y qué implicación tiene para este análisis?

**Respuesta 2.4:**

_Escribe aquí tu respuesta._
A): Hay muchas filas que son exactamente iguales en las 32 columnas, pero eso solo demuestra coincidencia de datos, no que necesariamente sean el mismo registro real.
El dataset tiene 119,390 reservas.
Se encontraron 31,994 filas sobrantes usando duplicated(keep='first'), aproximadamente 26.80% del total.
El problema no es igual por hotel:
City Hotel: 25,902 duplicados → 32.65% de sus reservas.
Resort Hotel: 6,092 → 15.21%.

b): duplicated() considera los NaN como parte de la comparación y en dado caso que exista Nan en ambos los toma como filas que son consideradas iguales.

Si el resto de las columnas también coincide, duplicated() puede marcar la segunda fila como duplicada. Esto es importante porque agent y company tienen muchos faltantes. Por ello, dos reservas distintas sin agente y sin compañía pueden terminar siendo consideradas idénticas si todos sus demás valores coinciden.

---
## Actividad 3. Impacto en el análisis (25 puntos)

### 3.1 Versión sin duplicados (5 pts)

Crea `hoteles_limpio` eliminando los duplicados exactos, conservando la primera aparición y **renumerando el índice**. Imprime el número de filas antes y después y cuántas se eliminaron.

In [72]:
# Escribe tu código aquí
# Elimina los duplicados exactos y renumera el índice
hoteles_limpio = hoteles.drop_duplicates(keep='first').reset_index(drop=True)

# Número de filas con el antes y después
filas_antes = len(hoteles)
filas_despues = len(hoteles_limpio)
filas_eliminadas = filas_antes - filas_despues

print(f"Filas antes: {filas_antes}")
print(f"Filas después: {filas_despues}")
print(f"Filas eliminadas: {filas_eliminadas}")


Filas antes: 119390
Filas después: 87396
Filas eliminadas: 31994


### 3.2 Métricas por hotel (8 pts)

Para `hoteles` y para `hoteles_limpio`, calcula por hotel con `groupby()`:

- Número de reservas (`.size()`).
- Tasa de cancelación (media de `is_canceled`).
- Tarifa diaria promedio (media de `adr`).

Nota: para calcular la media de varias columnas a la vez, selecciona una lista de columnas: `df.groupby('hotel')[['col1', 'col2']].mean()`.

Guarda las medias en dos `DataFrame`, `resumen_original` y `resumen_limpio`, con `hotel` como columna (usa `reset_index()`).

In [73]:
# Escribe tu código aquí
# Resumen del dataset original
resumen_original = (
    hoteles.groupby('hotel')
    .agg(
        reservas=('hotel', 'size'),
        tasa_cancelacion=('is_canceled', 'mean'),
        adr_promedio=('adr', 'mean')
    )
    .reset_index()
)

# Resumen del dataset limpio
resumen_limpio = (
    hoteles_limpio.groupby('hotel')
    .agg(
        reservas=('hotel', 'size'),
        tasa_cancelacion=('is_canceled', 'mean'),
        adr_promedio=('adr', 'mean')
    )
    .reset_index()
)

print("Resumen original:")
print(resumen_original)

print("\nResumen limpio:")
print(resumen_limpio)


Resumen original:
          hotel  reservas  tasa_cancelacion  adr_promedio
0    City Hotel     79330          0.417270    105.304465
1  Resort Hotel     40060          0.277634     94.952930

Resumen limpio:
          hotel  reservas  tasa_cancelacion  adr_promedio
0    City Hotel     53428          0.300386    110.985944
1  Resort Hotel     33968          0.234809     99.025346


### 3.3 Tabla comparativa (6 pts)

Agrega a cada resumen una columna `version` (`'original'` y `'sin duplicados'`) y únelos con `pd.concat()` en un `DataFrame` llamado `comparacion`, con índice renumerado y ordenado por `hotel`, de modo que las dos versiones de cada hotel queden juntas.

In [74]:
# Escribe tu código aquí

# Agrega la versión a cada resumen
resumen_original['version'] = 'original'
resumen_limpio['version'] = 'sin duplicados'

# Une ambos DataFrames
comparacion = pd.concat(
    [resumen_original, resumen_limpio],
    ignore_index=True
)

# Ordena por hotel para que las dos versiones queden juntas
comparacion = (
    comparacion
    .sort_values(by='hotel')
    .reset_index(drop=True)
)

print(comparacion)


          hotel  reservas  tasa_cancelacion  adr_promedio         version
0    City Hotel     79330          0.417270    105.304465        original
1    City Hotel     53428          0.300386    110.985944  sin duplicados
2  Resort Hotel     40060          0.277634     94.952930        original
3  Resort Hotel     33968          0.234809     99.025346  sin duplicados


### 3.4 Interpretación (6 pts)

a) ¿Qué métrica cambia más al eliminar duplicados? ¿En qué hotel el cambio es mayor? Explica por qué, relacionándolo con lo encontrado en la Actividad 2.

b) Si se entrenara un modelo para predecir `is_canceled` con los datos originales, ¿qué problemas causarían los duplicados en el entrenamiento y en la evaluación del modelo?

**Respuesta 3.4:**

_Escribe aquí tu respuesta._
a): La métrica que observé fue "reservas, tasa de cancelación y ADR promedio". Con los resultados de la Actividad 2, la diferencia más evidente debería estar en número de reservas, porque se eliminaron 31,994 filas, equivalentes al 26.80% del dataset.
B): Si se hace una división aleatoria después de conservar los duplicados, una fila y una copia idéntica podrían terminar una en entrenamiento y otra en prueba. El modelo podría encontrarse en evaluación con una observación prácticamente idéntica a una que ya vio durante el entrenamiento (lo que se conoce como data leakage).


---
## Actividad 4. Tratamiento y verificación (20 puntos)

### 4.1 Verificación (5 pts)

Verifica sobre `hoteles_limpio` que:

- No quedan duplicados exactos.
- No hay índices repetidos y el índice va de 0 a `len(hoteles_limpio) - 1`.

In [75]:
# Escribe tu código aquí
# 1. Verifica que no existan duplicados exactos
sin_duplicados = not hoteles_limpio.duplicated().any()

# 2. Verifica que no haya índices repetidos
indice_sin_repetir = hoteles_limpio.index.is_unique

# 3. Verifica que el índice sea exactamente 0, 1, 2, ..., len-1
indice_correcto = hoteles_limpio.index.equals(
    pd.RangeIndex(start=0, stop=len(hoteles_limpio))
)

print(f"¿Quedan duplicados exactos?: {sin_duplicados}")
print(f"¿El índice no tiene valores repetidos?: {indice_sin_repetir}")
print(f"¿El índice va de 0 a len(hoteles_limpio)-1?: {indice_correcto}")


¿Quedan duplicados exactos?: True
¿El índice no tiene valores repetidos?: True
¿El índice va de 0 a len(hoteles_limpio)-1?: True


### 4.2 Duplicados parciales (10 pts)

La columna `reservation_status_date` registra la fecha del último cambio de estado. Detecta las filas de `hoteles_limpio` que **coinciden en todas las columnas excepto `reservation_status_date`**.

1. Cuenta cuántas filas sobrantes hay con ese criterio.
2. Muestra algunas de esas filas ordenadas, con las columnas `hotel`, `lead_time`, `arrival_date_year`, `arrival_date_month`, `arrival_date_day_of_month`, `adr`, `reservation_status` y `reservation_status_date`.
3. Muestra la distribución de `reservation_status` en esas filas.

Después responde: ¿qué podrían representar estas filas? ¿Las eliminarías? Justifica.

In [76]:
# Escribe tu código aquí


**Respuesta 4.2:**

_Escribe aquí tu respuesta._

### 4.3 Registro de decisión (5 pts)

Redacta, en una celda de texto, el registro de la limpieza realizada, como lo documentarías en un proyecto real. Incluye:

- **Criterio** de duplicado utilizado (columnas comparadas).
- **Regla de conservación** (`keep`) y por qué es o no relevante en este caso.
- **Resultado** (filas antes, después y eliminadas).
- **Supuestos y limitaciones**.

**Respuesta 4.3:**

_Escribe aquí tu respuesta._

---
## Reflexión final (10 puntos)

Responde brevemente, con base en tus resultados:

1. Describe una situación en la que **no** convendría eliminar los duplicados exactos de este dataset.
2. ¿Qué información adicional (columnas) le pedirías al proveedor de los datos para tomar una decisión más segura? ¿Por qué?
3. En tus propias palabras, ¿por qué `duplicated().sum() == 0` no garantiza que un dataset esté libre de duplicados?

**Respuesta final:**

_Escribe aquí tu respuesta._
1): Creo que no conviene eliminar datos si dos reservas diferentes pueden tener exactamente los mismos datos registrados. En este dataset, al no existir un identificador de reserva, una coincidencia exacta no demuestra que sea el mismo registro. Por otro lado encontramos 31,994 filas sobrantes (26.80%), por lo que eliminarlas tiene un impacto considerable.
2): Sería perdirle al proveedor el ID único de reserva, porque permitiría determinar directamente si dos filas representan la misma reserva. La segunda información que pediría es quién realizó el llenado de la información para preguntar directamente el motivo de algun dato duplicado de un cliente.
3): Porque duplicated() solo detecta filas que son iguales según las columnas que se están comparando. Dos registros pueden representar la misma reserva pero tener alguna diferencia

---
**Fin de la actividad.** Antes de entregar: **Reiniciar y ejecutar todo**, verificar que no haya errores, guardar con las salidas visibles y comprobar que el enlace abre sin iniciar sesión.